File: Samples/Verbum/how-to-emulate-the-ms-word-information-status-bar-when-using-richeditcontrol-t467150-16.1.9-/CS/WindowsFormsApplication1/CustomLayoutVisitor.cs
using DevExpress.XtraRichEdit.API.Layout;
using DevExpress.XtraRichEdit.API.Native;
using DevExpress.XtraRichEdit.Layout;
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Linq;
using System.Reflection;
using System.Text;

namespace WindowsFormsApplication1
{
    public class StaticsticsVisitor : DocumentVisitorBase
    {
        readonly StringBuilder buffer;
        
        public int WordCount { get; private set; }
        protected StringBuilder Buffer { get { return buffer; } }

        public StaticsticsVisitor()
        {
            WordCount = 0;
            this.buffer = new StringBuilder();
        }

        public override void Visit(DocumentText text)
        {
            Buffer.Append(text.Text);
        }

        public override void Visit(DocumentParagraphEnd paragraphEnd)
        {
            FinishParagraph();
        }

        void FinishParagraph()
        {
            string text = Buffer.ToString();
            this.WordCount += text.Split(new char[] { ' ', '.', '!', '?' }, StringSplitOptions.RemoveEmptyEntries).Length;
            Buffer.Length = 0;
        }
    }

    public class CustomLayoutVisitor : LayoutVisitor
    {
        Document document;

        public CustomLayoutVisitor(Document doc)
        {
            this.document = doc;
            RowIndex = 0;
            ColIndex = 0;
            IsFound = false;
        }

        protected override void VisitRow(LayoutRow row)
        {
            if (!IsFound)
            {
                RowIndex++;
                if (row.Range.Contains(document.CaretPosition.ToInt()))
                {
                    IsFound = true;
                    ColIndex = document.CaretPosition.ToInt() - row.Range.Start;
                }
            }
            base.VisitRow(row);
        }

        public void Reset()
        {
            RowIndex = 0;
            ColIndex = 0;
        }

        public int ColIndex { get; private set; }
        public int RowIndex { get; private set; }
        public bool IsFound { get; private set; }
    }
}

File: Samples/Verbum/how-to-emulate-the-ms-word-information-status-bar-when-using-richeditcontrol-t467150-16.1.9-/VB/WindowsFormsApplication1/CustomLayoutVisitor.vb
Imports DevExpress.XtraRichEdit.API.Layout
Imports DevExpress.XtraRichEdit.API.Native
Imports DevExpress.XtraRichEdit.Layout
Imports System
Imports System.Collections.Generic
Imports System.Drawing
Imports System.Linq
Imports System.Reflection
Imports System.Text

Namespace WindowsFormsApplication1
    Public Class StaticsticsVisitor
        Inherits DocumentVisitorBase

        Private ReadOnly _buffer As StringBuilder

        Public Sub New()
            WordCount = 0
            Me._buffer = New StringBuilder()
        End Sub

        Public Overrides Sub Visit(ByVal text As DocumentText)
            _buffer.Append(text.Text)
        End Sub

        Public Overrides Sub Visit(ByVal paragraphEnd As DocumentParagraphEnd)
            FinishParagraph()
        End Sub

        Private Sub FinishParagraph()
            Dim text As String = _buffer.ToString()
            Me.WordCount += text.Split(New Char() { " "c, "."c, "!"c, "?"c }, StringSplitOptions.RemoveEmptyEntries).Length
            _buffer.Length = 0
        End Sub

        Private privateWordCount As Integer
        Public Property WordCount() As Integer
            Get
                Return privateWordCount
            End Get
            Private Set(ByVal value As Integer)
                privateWordCount = value
            End Set
        End Property
        Protected ReadOnly Property Buffer() As StringBuilder
            Get
                Return _buffer
            End Get
        End Property
    End Class

    Public Class CustomLayoutVisitor
        Inherits LayoutVisitor

        Private document As Document

        Public Sub New(ByVal doc As Document)
            Me.document = doc
            RowIndex = 0
            ColIndex = 0
            IsFound = False
        End Sub

        Protected Overrides Sub VisitRow(ByVal row As LayoutRow)
            If Not IsFound Then
                RowIndex += 1
                If row.Range.Contains(document.CaretPosition.ToInt()) Then
                    IsFound = True
                    ColIndex = document.CaretPosition.ToInt() - row.Range.Start
                End If
            End If
            MyBase.VisitRow(row)
        End Sub

        Public Sub Reset()
            RowIndex = 0
            ColIndex = 0
        End Sub

        Private privateColIndex As Integer
        Public Property ColIndex() As Integer
            Get
                Return privateColIndex
            End Get
            Private Set(ByVal value As Integer)
                privateColIndex = value
            End Set
        End Property
        Private privateRowIndex As Integer
        Public Property RowIndex() As Integer
            Get
                Return privateRowIndex
            End Get
            Private Set(ByVal value As Integer)
                privateRowIndex = value
            End Set
        End Property
        Private privateIsFound As Boolean
        Public Property IsFound() As Boolean
            Get
                Return privateIsFound
            End Get
            Private Set(ByVal value As Boolean)
                privateIsFound = value
            End Set
        End Property
    End Class
End Namespace
